site/docs/design-docs/018-kinesis-connectors.md

```diff
@@ -0,0 +1,476 @@
+the MD5 hash function to map these strings to 128-bit integer values.
+The universe of these values is thus the [0 .. 2^128) range. Each
+Kinesis shard has a continuous chunk of this range assigned to it,
+called the shard's hash range. The stream assigns a record to a shard if
+the record's partition key hashes into the shard's range.
+
+In the Jet Kinesis source, we use similar logic for assigning shards to
+source instances. Each of our sources gets a part of the hash range
+assigned to it. We say that a source owns a specific shard if and only
+if the shard's hash range's starting point is inside the source's hash
+range. Any similar range matching logic would work, as long as it's
+non-ambiguous.
+
+### Discovery
+
+Reading records from shards assigned to them is only part of the
+responsibility of sources. Sources also need a way to discover currently
+active shards in the stream to take responsibility for them. Moreover,
+this discovery process can't just happen once, on start-up, because
+streams are dynamic, shards can be closed, and new shards can pop-up at
+any time. For details, see the [resharding section](#resharding).
+
+Continuously monitoring the set of active shards in the stream is the
+responsibility of **one** of the local source instances in each Jet
+cluster member. This is an optimization. If all sources would run the
+discovery, they would still obtain the same data, just with a multiplied
+effort and cost. Monitoring means continuously polling the stream for
+the list of all shards in it.
+
+Monitoring needs to take care not to cross the rate limit imposed by
+Kinesis on this operation. For details, see the [quotas
+section](#quotas).
+
+### Resharding
+
+Kinesis supports resharding, which lets you adjust the number of shards
+in your stream to adapt to changes in data flow rate through the stream.
+(Amazon charges on a per-shard basis, that's why it's desirable to have
+the smallest amount of shards possible.)
+
+There are two types of resharding operations: shard **split** and shard
+**merge**. In a shard split, you divide a single shard into two adjacent
+shards. In a shard merge, you combine two adjacent shards into a single
+shard. By "adjacent", we mean that one's hash range starts where the
+other one's ends.
+
+Splitting increases the number of shards in your stream and therefore
+increases the data capacity (and cost) of the stream. Similarly, merging
+reduces the number of shards in your stream and therefore decreases the
+data capacity (and cost).
+
+Resharding is always pairwise in the sense that you cannot split into
+more than two shards in a single operation, and you cannot merge more
+than two shards in a single operation. The shard or pair of shards that
+the resharding operation acts on are called parent shards. The shard or
+pair of shards that result from the resharding operation are called
+child shards.
+
+When child shards, resulting from a split or merge, activate, their
```

Comment:
I don't know. Testing has revealed that if I make sure to read all data from the parents, until reading returns EOF and then switching to the children and reading then I will not lose data, but what exactly happens, if the switch is clean and well defined, I don't know. The documentation doesn't say, so even if I dig up the details from the code, I can't rely on it, because it might change. I wouldn't be surprised if it wouldn't be well defined, because I've just witnessed for example a series of "list shard" operations (on the real back-end) which at some point showed me child shards and then went back to an obsolete state with only the parent shards and then recent again.... So my current impression of Kinesis is that if you know what's good for you, then you don't trigger a resharding while data is flowing through....